### ISSUE TITLE
Odysee (odysee.com) support

### ISSUE THREAD

**Issue by fedelibre** [bravepipeproject/BravePipeExtractor#16]
Thanks for this fork!

I'd be glad to be able to see Odyssey videos on BraveNewPipe.

**Comment by strider72** [comment]
URL is https://odysee.com/ not odyssee

**Comment by herrdeh** [comment]
That's the one I'm really missing!

**Comment by fedelibre** [comment]
> URL is https://odysee.com/ not odyssee

thanks, title updated

**Comment by strider72** [comment]
I guess more specifically this would be LBRY support...?

**Comment by GH462516** [comment]
Yes Odysee/LBRY definitely needed. Odysee seems to have more content providers than Rumble and Bitchute. Having all 3 platforms would make this the perfect app

**Comment by evermind-zz** [comment]
Just for reference https://github.com/TeamNewPipe/NewPipeExtractor/issues/434

**Comment by Neurognostic** [comment]
See [LBRYFoundation/Awesome-LBRY](https://github.com/LBRYFoundation/Awesome-LBRY) for various LBRY client implementations and tools.

**Comment by Fehertukor** [comment]
I would love to use this function too.
Please if you have time integrate it.

Thanks!

**Comment by wepache** [comment]
I would love LBRY/Odysee support on this one also. A lot of nonpolitical channels prefer Odysee and I do follow many on Odysee.

**Comment by Mr-Bajs** [comment]
I would be most welcome if Odysee/Lbry was added.

**Comment by evermind-zz** [comment]
I noticed that during looking into the official odysee-android app that the code that could be relevant here for the extractor is mixed with android code. Reusing that code is not easy as NewPipeExtractor should not depend on android libraries. I do not understand why they did not abstract it. Does anyone know why? Is there any third party java based LBRY api?

The only thing I found (just a quick search) was:
some information about the API. This information are third party and not from LBRY
https://librarian.codeberg.page/docs/

**Comment by nicoursi** [comment]
Could this be useful to you?

https://github.com/TeamNewPipe/NewPipeExtractor/issues/434#issuecomment-719995735

**Comment by nicoursi** [comment]
> I noticed that during looking into the official odysee-android app that the code that could be relevant here for the extractor is mixed with android code. Reusing that code is not easy as NewPipeExtractor should not depend on android libraries. I do not understand why they did not abstract it. Does anyone know why? Is there any third party java based LBRY api?
> 
> The only thing I found (just a quick search) was: some information about the API. This information are third party and not from LBRY https://librarian.codeberg.page/docs/

As far as i know the official LBRY APIs are in python. So probably the native code you'retalking about coil just be a wrapper in C?

> My preferred way to integrate the two would use RPC. [XML RPC](http://en.wikipedia.org/wiki/XML-RPC) is not a bad choice here, if you have moderate amounts of data. It is pretty well supported — Python has it in its standard library. Java libraries are also easy to find. Now depending on your setup either Java or Python part would be a server accepting connection from other language.

> A less popular but worth considering alternative way to do RPCs is Google protobuffers, which have 2/3 of support for [nice rpc](http://code.google.com/apis/protocolbuffers/docs/reference/java-generated.html#service). You just need to provide your transport layer. Not that much work and the convenience of writing is reasonable.

> Another option is to write a C wrapper around that pieces of Python functionality that you need to expose to Java and use it via JVM native plugins. You can ease the pain by going with SWIG [SWIG](http://www.swig.org/).

> Essentially in your case it works like that:

> Create a SWIG interface for all method calls from Java to C++.
> Create C/C++ code that will receive your calls and internally call python interpreter with right params.
Convert response you get from python and send it via swig back to your Java code.
> This solution is fairly complex, a bit of an overkill in most cases. Still it is worth doing if you (for some reason) cannot afford RPCs. RPC still would be my preferred choice, though.

Source: https://stackoverflow.com/questions/1119696/Java-python-integration
Python LBRY APIs: https://github.com/lbryio/lbry-sdk/

**Comment by nicoursi** [comment]
This is a port in java but it is quiet old. You might want to give it a try:. https://github.com/hdost/lbry-java

**Comment by nicoursi** [comment]
Did you have a look at this app? It might be what you're looking for: https://github.com/lbryio/lbry-fdroid

**Comment by evermind-zz** [comment]
@nicoursi 
thx for your reply. Indeed I looked into the official odysee app and had the same feelings than you about the abstraction.

Nevertheless I tried to split the code and/or ripped chunks of code into a java only library. It's ugly but we will see.

- With this java library the original (my fork) odysee app is running fine.
-  I've also started to use this library for integrating odysee into the extractor.  I will release the code as soon as I find some time to do so.. So far only stream extracting part is nearly finished. Hopefully the stream extracting code will soon have a working unit test.

**Comment by nicoursi** [comment]
That's cool you already managed to extract good code.

I'd suggest though to have  also a look at the former lbry client for android, the api seem to be 100% in java. Maybe it is easier to incorporate it.  

https://github.com/lbryio/lbry-fdroid/tree/master/app/src/main/java/io/lbry/browser/utils

**Comment by JERW86** [comment]
@evermind-zz 
Thank you for adding and maintaining more video platforms to this app. Here is this Odysee plugin code. I hope it can reduce the work involved in making a plugin for Odysee. It is from another open source video app called Grayjay. Maybe it has some extractor points someone can use to make and add an Odysee plugin?

App website: https://grayjay.app/ 

Odyssey Plugin Code:
https://plugins.grayjay.app/Odysee/OdyseeScript.js

Plugin documentation:  https://gitlab.futo.org/videostreaming/grayjay/-/blob/master/plugin-development.md

**Comment by DDQW** [comment]
i would also like to get lbry in. Any news?

**Comment by ghost** [comment]
This would be awesome!  It would be nice if possible that if a youtube video has been detected as on Odysee/LBRY it can automatically load the Odysee video. Like this extension does:
https://addons.mozilla.org/en-GB/firefox/addon/watch-on-odysee/

**Comment by clearins1ght** [comment]
Bravepipe is amazing and so is Odysee. This feature would therefore be awesome.

**Comment by evermind-zz** [comment]
> [@nicoursi](https://github.com/nicoursi) thx for your reply. Indeed I looked into the official odysee app and had the same feelings than you about the abstraction.
> 
> Nevertheless I tried to split the code and/or ripped chunks of code into a java only library. It's ugly but we will see.
> 
> * With this java library the original (my fork) odysee app is running fine.
> * I've also started to use this library for integrating odysee into the extractor.  I will release the code as soon as I find some time to do so.. So far only stream extracting part is nearly finished. Hopefully the stream extracting code will soon have a working unit test.

Long overdue, but here is my code from back then trying to seperate the android bits from the odysee-android app and creating a odysee-api library 

Here basically I have a fork of OdyseeTeam/odysee-android (https://github.com/evermind-zz/odysee-android) and the extracted api thereof in another repository (https://github.com/evermind-zz/odysee-api) that is included in there.


I have no idea if the app is still working at all, but here it is if someone wants to have a look into it let me know

**Comment by slimhk45** [comment]
+1 Odysee is the main decentralized alternative to YouTube for creators. An essential.
